### PR TITLE
chore(utils): make cache utils `callonce()` compatible with py 3.11

### PR DESCRIPTION
**Pull request by Yun-Kim**
## Description
#4302 added a `callonce()` decorator as a cache utility function. However, this imported `inspect.getargspec()` was removed as of Python 3.11. This was fixed by importing `inspect.getfullargspec()`, which is the recommended replacement for `getargspec()`. Note that `getargspec()` is still used in Python 2.

## Checklist
- [ ] Title must conform to [conventional commit](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional).
- [ ] Add additional sections for `feat` and `fix` pull requests.
- [ ] Ensure tests are passing for affected code.
- [ ] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Reviewer Checklist
- [ ] Title is accurate.
- [ ] Description motivates each change.
- [ ] No unnecessary changes were introduced in this PR.
- [ ] PR cannot be broken up into smaller PRs.
- [ ] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.
- [ ] Add to milestone.

### CODE REVIEW

**Review by mabdinur — APPROVED** [review]

File: ddtrace/internal/utils/cache.py [hunk 1]
```diff
@@ -107,10 +113,22 @@ def cached_wrapper(f):
...
-
-    argspec = getargspec(f)
-    if argspec.args or argspec.varargs or argspec.keywords or argspec.defaults or isgeneratorfunction(f):
-        raise ValueError("The callonce decorator can only be applied to functions with no arguments")
+    if PY2:
+        argspec = getargspec(f)
+        if argspec.args or argspec.varargs or argspec.keywords or argspec.defaults or isgeneratorfunction(f):
+            raise ValueError("The callonce decorator can only be applied to functions with no arguments")
+    else:
+        argspec = getfullargspec(f)
```
Comment (brettlangdon) [inline ddtrace/internal/utils/cache.py:121]: are we able to have a backported/compatible version of `getfullargspec` in `ddtrace/internal/compat.py` ?

Comment (Yun-Kim) [reply ddtrace/internal/utils/cache.py:121]: @brettlangdon Is a backport necessary? This should only be a problem for Python 3.11 onwards.

Comment (Yun-Kim) [reply ddtrace/internal/utils/cache.py]: Moved this to `ddtrace/internal/compat.py`.

File: ddtrace/internal/compat.py [hunk 2]
```diff
@@ -84,6 +85,27 @@
 else:
     pattern_type = re._pattern_type  # type: ignore[misc,attr-defined]
 
+try:
+    from inspect import getargspec as getfullargspec
+
+    def args_provided(f, argspec):
```
Comment (P403n1x87) [inline ddtrace/internal/compat.py:91]: Whilst the logic is correct, I think `args_provided` is a misnomer for this helper as `isgeneratorfunction` is not a check on arguments. IMO something more appropriate is `is_argless_function`/`is_void_function` ... .

**Comment by P403n1x87** [comment]
🙌  thanks for catching, and taking care of this!

**Review by majorgreys — APPROVED** [review]

**Review by ZStriker19 — APPROVED** [review]

**Review by jd — APPROVED** [review]

**Pull request merged**